src/main/kotlin/cacophony/codegen/instructions/Instruction.kt

```diff
@@ -25,10 +25,11 @@ data class MemoryAddress(val base: Register, val index: Register?, val scale: In
 
 interface Instruction {
     val registersRead: Set<Register>
-
     val registersWritten: Set<Register>
 
     fun toAsm(hardwareRegisterMapping: HardwareRegisterMapping): String
+
+    fun substituteVirtualRegisters(map: Map<Register, Register>): Instruction
```

Comment:
This is only used with VirtualRegisters, but I think that `substituteRegisters` would be a better name given the function parameters.